wrappers/pkcs11/pkcs11.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2024 OpenBao a Series of LF Projects, LLC
+// SPDX-License-Identifier: MPL-2.0
+
+package pkcs11
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	wrapping "github.com/openbao/go-kms-wrapping/v2"
+)
+
+// Wrapper is a Wrapper that uses PKCS11
+type Wrapper struct {
+	client       pkcs11ClientEncryptor
+	keyId        string
+	currentKeyId *atomic.Value
+}
+
+// Ensure that we are implementing Wrapper
+var _ wrapping.Wrapper = (*Wrapper)(nil)
+
+// NewWrapper creates a new PKCS11 Wrapper
+func NewWrapper() *Wrapper {
+	k := &Wrapper{
+		currentKeyId: new(atomic.Value),
+	}
+	k.currentKeyId.Store("")
+	return k
+}
+
+// Init is called during core.Initialize
+func (k *Wrapper) Init(_ context.Context) error {
+	return nil
+}
+
+// Finalize is called during shutdown
+func (k *Wrapper) Finalize(_ context.Context) error {
+	k.client.Close()
+	return nil
+}
+
+// SetConfig processes the config info from the server config
+func (k *Wrapper) SetConfig(_ context.Context, opt ...wrapping.Option) (*wrapping.WrapperConfig, error) {
+	opts, err := getOpts(opt...)
+	if err != nil {
+		return nil, err
+	}
+
+	client, wrapConfig, err := newPkcs11Client(opts)
+	if err != nil {
+		return nil, err
+	}
+	k.client = client
+	k.keyId = client.GetCurrentKey().String()
+
+	// Send a value to test the wrapper and to set the current key id
+	if _, err := k.Encrypt(context.Background(), []byte("a")); err != nil {
```

Comment:
I think I'll remove the encryption test here. The PKCS11 client already test the session.
Regarding the mechanisms, of course AEAD are better.

Something currently wrong in the implementation is the parameter `hmac_mechanism` is not handled and should be required for CBC and CBC PAD.
At first, I was thinking that SoftHSM2 did not support AES GCM, but it's just the documentation which is not up-to-date.
There is still the issue with YubiHSM2 but yes, it might be interesting to review the list of supported mechanisms !

Reply:
I'm not inclined to support hmac parameter and just require an AEAD mech.

Reply:
Yes, of course.
It just like you wrote 3 days ago: to support encrypt-then-mac it requires to add the HMAC mechanism parameter.